File: src/main/scala/com/nawforce/common/documents/Location.scala
/*
 [The "BSD licence"]
 Copyright (c) 2019 Kevin Jones
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.nawforce.common.documents

import com.nawforce.common.path.PathLike

abstract class Location(val path: PathLike, val line: Int) {
  def startPosition: (Int, Int) = (line, 0)
  def endPosition: (Int, Int) = (line+1, 0)
  def displayPosition: String
  def asJSON: String
}

case class LineLocation(_path: PathLike, _line: Int) extends Location(_path, _line) {
  override def displayPosition: String = s"line $line"
  override def asJSON: String = s""""start": {"line": $line}"""
}

case class LineRangeLocation(_path: PathLike, start: Int, end: Int) extends Location(_path, start) {
  override def endPosition: (Int, Int) = (end+1, 0)
  override def displayPosition: String = s"line $start to $end"
  override def asJSON: String = s""""start": {"line": $start}, "end": {"line": $end}"""
}

case class Position(line: Int, offset: Int) {
  def getPosition: (Int, Int) = (line, offset)
  def displayPosition: String = s"line $line at $offset"
  def asJSON: String = s"""{"line": $line, "offset": $offset}"""

  def adjust(lineOffset: Int, positionOffset: Int): Position = {
    if (lineOffset == 0 && positionOffset == 0) {
      return this
    }

    if (line == 1)
      Position(line + lineOffset, offset + positionOffset)
    else
      Position(line + lineOffset, offset)
  }
}

case class TextRange(start: Position, end: Position) {
  def startPosition: (Int, Int) = start.getPosition
  def displayPosition: String = {
    if (start.line == end.line)
      s"line ${start.line} at ${start.offset}-${end.offset}"
    else
      s"line ${start.line} to ${end.line}"
  }

  def asJSON: String =
    s""""start": ${start.asJSON}, "end": ${end.asJSON}"""

  def adjust(lineOffset: Int, positionOffset: Int) : TextRange = {
    if (lineOffset == 0 && positionOffset == 0) {
      return this
    }

    TextRange(
      start.adjust(lineOffset, positionOffset),
      end.adjust(lineOffset, positionOffset)
    )
  }
}

case class PointLocation(_path: PathLike, start: Position) extends Location(_path, start.line) {
  override def startPosition: (Int, Int) = start.getPosition
  override def endPosition: (Int, Int) = (startPosition._1+1, 0)
  override def displayPosition: String = start.displayPosition
  override def asJSON: String = start.asJSON
}

case class RangeLocation(_path: PathLike, start: Position, end: Position) extends Location(_path, start.line) {
  override def startPosition: (Int, Int) = start.getPosition
  override def endPosition: (Int, Int) = end.getPosition
  override def displayPosition: String = TextRange(start, end).displayPosition
  override def asJSON: String = TextRange(start, end).asJSON
}

object TextRange {
  val empty: TextRange = TextRange(Position(0, 0), Position(0, 0))

  def apply(line: Int): TextRange = {
    TextRange(
      Position(line, 0),
      Position(line+1, 0)
    )
  }
}

object RangeLocation {
  def apply(path: PathLike, range: TextRange): RangeLocation = {
    RangeLocation(path, range.start, range.end)
  }
}

File: src/main/java/com/nawforce/platform/SObjects/Location.java
/*
 [The "BSD licence"]
 Copyright (c) 2019 Kevin Jones
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package com.nawforce.platform.SObjects;

import com.nawforce.platform.Internal.SObjectFields$;
import com.nawforce.platform.Internal.SObjectType$;
import com.nawforce.platform.System.Boolean;
import com.nawforce.platform.System.Integer;
import com.nawforce.platform.System.String;
import com.nawforce.platform.System.*;

@SuppressWarnings("unused")
public class Location extends SObject {
	public static SObjectType$<Location> SObjectType;
	public static SObjectFields$<Location> Fields;
	public Date CloseDate;
	public Date ConstructionEndDate;
	public Date ConstructionStartDate;
	public Id CreatedById;
	public User CreatedBy;
	public Datetime CreatedDate;
	public String Description;
	public String DrivingDirections;
	public Id Id;
	public Boolean IsDeleted;
	public Boolean IsInventoryLocation;
	public Boolean IsMobile;
	public Id LastModifiedById;
	public User LastModifiedBy;
	public Datetime LastModifiedDate;
	public Datetime LastReferencedDate;
	public Datetime LastViewedDate;
	public Decimal Latitude;
	public Location Location;
	public Integer LocationLevel;
	public String LocationType;
	public Decimal Longitude;
	public String Name;
	public Date OpenDate;
	public Id OwnerId;
	public Group Owner;
	public Id ParentLocationId;
	public Location ParentLocation;
	public Date PossessionDate;
	public Date RemodelEndDate;
	public Date RemodelStartDate;
	public Id RootLocationId;
	public Location RootLocation;
	public Datetime SystemModstamp;
	public String TimeZone;
	public Id VisitorAddressId;
	public Address VisitorAddress;

	public ActivityHistory[] ActivityHistories;
	public Address[] Addresses;
	public Asset[] Assets;
	public AssociatedLocation[] AssociatedLocations;
	public AttachedContentDocument[] AttachedContentDocuments;
	public Attachment[] Attachments;
	public Location[] ChildLocations;
	public CombinedAttachment[] CombinedAttachments;
	public ContentDocumentLink[] ContentDocumentLinks;
	public Shipment[] DestinationLocationShipments;
	public ProductRequestLineItem[] DestinationProductRequestLineItems;
	public ProductRequest[] DestinationProductRequests;
	public ProductTransfer[] DestinationProductTransfers;
	public ReturnOrderLineItem[] DestinationReturnOrderLineItems;
	public ReturnOrder[] DestinationReturnOrders;
	public EmailMessage[] Emails;
	public Event[] Events;
	public EntitySubscription[] FeedSubscriptionsForEntity;
	public LocationFeed[] Feeds;
	public LocationHistory[] Histories;
	public MaintenancePlan[] MaintenancePlans;
	public Note[] Notes;
	public NoteAndAttachment[] NotesAndAttachments;
	public OpenActivity[] OpenActivities;
	public ProcessInstance[] ProcessInstances;
	public ProcessInstanceHistory[] ProcessSteps;
	public ProductItem[] ProductItems;
	public RecordActionHistory[] RecordActionHistories;
	public RecordAction[] RecordActions;
	public ServiceResource[] ServiceResources;
	public ServiceTerritoryLocation[] ServiceTerritories;
	public LocationShare[] Shares;
	public Shipment[] SourceLocationShipments;
	public ProductRequestLineItem[] SourceProductRequestLineItems;
	public ProductRequest[] SourceProductRequests;
	public ProductTransfer[] SourceProductTransfers;
	public ReturnOrderLineItem[] SourceReturnOrderLineItems;
	public ReturnOrder[] SourceReturnOrders;
	public Task[] Tasks;
	public WorkOrderLineItem[] WorkOrderLineItems;
	public WorkOrder[] WorkOrders;

	public Location clone$() {throw new java.lang.UnsupportedOperationException();}
	public Location clone$(Boolean preserveId) {throw new java.lang.UnsupportedOperationException();}
	public Location clone$(Boolean preserveId, Boolean isDeepClone) {throw new java.lang.UnsupportedOperationException();}
	public Location clone$(Boolean preserveId, Boolean isDeepClone, Boolean preserveReadonlyTimestamps) {throw new java.lang.UnsupportedOperationException();}
	public Location clone$(Boolean preserveId, Boolean isDeepClone, Boolean preserveReadonlyTimestamps, Boolean preserveAutonumber) {throw new java.lang.UnsupportedOperationException();}
}
